first draft for ebs csi driver (#720)
giantswarm/docs: src/content/guides/using-persistent-volumes-with-csi-driver-on-aws/index.md

---

> title: "Using Persistent Volumes with EBS CSI driver on AWS"

Reviewer: ```suggestion
title: Using persistent volumes with the EBS CSI driver on AWS
```

- Quotes not needed
- We use sentence case for headlines (since very recently)

> description: Tutorial on how to use dynamically provisioned Persistent Volumes with EBS CSI driver on a cluster running on Amazon Web Services"

Reviewer: ```suggestion
description: Tutorial on how to use dynamically provisioned persistent volumes with the EBS CSI driver on a cluster running on Amazon Web Services (AWS)."
```

type: page
weight: 50
tags: ["tutorial"]
user_questions:
  - How do I install the EBS CSI driver?
  - How do I provision an EBS instance on AWS?
  - How do I test the EBS CSI storage class?

>   - How do I use EBS CSI Volumes?

Reviewer: ```suggestion
  - How do I use EBS CSI volumes?
```

owner:
  - https://github.com/orgs/giantswarm/teams/team-firecracker
---

> # Using Persistent Volumes on AWS with Container Storage Interface (CSI)

Reviewer: Ideally this headline should be identical with the title in front matter.

> CSI has been promoted to GA in Kubernetes v1.13 and is becoming the standard to replace the current Kubernetes `in-tree` storage plugin to handle Volumes for different provider.

Reviewer: Ideally we introduce non-common abbreviations (or those where we are not certain if everybody understands them) in text by giving the long form first and the abbreviation in parentheses.

```suggestion
The Container Storage Interface (CSI) has been promoted to general availability (GA) in Kubernetes v1.13 and is becoming the standard to replace the current Kubernetes `in-tree` storage plugin to handle volumes for different providers.
```

> The `aws-ebs-csi-driver` provides a CSI interface to manage the lifecycle of Amazon EBS Volumes.

Author: ```suggestion
The `aws-ebs-csi-driver-app` provides a CSI interface to manage the lifecycle of Amazon EBS Volumes.
```

> The `aws-ebs-csi-driver-app` provides a CSI interface to manage the lifecycle of Amazon EBS Volumes.

Author: The app is for installing the ebs csi driver, without this app you cannot create volumes via CSI. See https://github.com/giantswarm/aws-ebs-csi-driver-app.

> The `aws-ebs-csi-driver-app` provides a CSI interface to manage the lifecycle of Amazon EBS Volumes.

Reviewer: Yeah, I figured that out. I wanted to point out questions the users will ask themselves. Anyway, improved. Thanks!

> ## Installing EBS CSI driver

Reviewer: ```suggestion
## Installing the EBS CSI driver
```

To install the EBS CSI driver you will need to follow these steps:

> 1. Access Giant Swarm web UI and select the cluster on which you want to install the EBS CSI driver.

Reviewer: ```suggestion
1. Access the [web interface](/reference/web-interface/) and select the cluster on which you want to install the EBS CSI driver.
```

> 2. Open the _Giant Swarm Playground_ catalog.
> 3. Write `aws-ebs-csi-driver-app` in the search bar.
> 4. Click the _Configure & Install_ button, select the cluster and it will be installed.

Reviewer: ```suggestion
2. Open the Apps tab.
3. Click the Install App button
4. Select the Giant Swarm Playground catalog.
5. Select the App named aws-ebs-csi-driver-app.
6. Click the Configure & Install button. Make sure that the correct cluster is selected.
7. Click the Install App button.
```

Adding some intermediate steps. No need for using search here, as the app is in the first row.

> ## Storage Classes

Reviewer: ```suggestion
## Storage classes
```

> Your Kubernetes cluster will have a Storage Class `ebs-csi` deployed, once you installed the `aws-ebs-csi-driver` from _Giant Swarm Playground_ catalog.

Author: ```suggestion
Your Kubernetes cluster will have a Storage Class `ebs-csi` deployed, once you installed the `aws-ebs-csi-driver-app` from _Giant Swarm Playground_ catalog.
```

> Your Kubernetes cluster will have a Storage Class `ebs-csi` deployed, once you installed the `aws-ebs-csi-driver-app` from _Giant Swarm Playground_ catalog.

Reviewer: ```suggestion
Your Kubernetes cluster will have a storage class `ebs-csi` deployed, once you installed the `aws-ebs-csi-driver-app` from _Giant Swarm Playground_ catalog.
```

I would turn that around. "Once you have ... installed, your cluster ... "

> The storage class supports volume encryption and resizing by default.

Author: ```suggestion
The storage class offers volume encryption and allows resizing by default.
```

## Creating Persistent Volumes

> The most straight forward way to create a Persistent Volume is to create a [Persistent Volume Claim object](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims), which will automatically create a corresponding Persistent Volume (PV) for you.

Reviewer: ```suggestion
The most straight forward way to create a persistent volume is to create a [PersistentVolumeClaim](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims) resource, which will automatically create a corresponding `PersistentVolume` (PV) resource for you.
```

> Alternatively, to be able to set more specific parameters on your PV you can first create a [Persistent Volume object](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistent-volumes) and then claim that PV using a Persistent Volume Claim (PVC).

Reviewer: ```suggestion
Alternatively, to be able to set more specific parameters on your PV, you can first create a [PersistentVolume](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistent-volumes) resource and then claim that PV using a `PersistentVolumeClaim` (PVC).
```

> Under the hood, the EBS CSI driver will take care that a corresponding EBS Volume with the correct parameters is created.

Reviewer: ```suggestion
Under the hood, the EBS CSI driver will take care that a corresponding EBS volume with the correct parameters is created.
```

> The EBS Volume and its data will persist as long as the corresponding PV resource exists. Deleting the resource will also delete the corresponding EBS volume, which means that all stored data will be lost at that point.

Reviewer: ```suggestion
The EBS volume and its data will persist as long as the corresponding PV resource exists. Deleting the PV resource will also delete the corresponding EBS volume, which means that all stored data will be lost at that point.
```

> ## Using Persistent Volumes in a Pod

Reviewer: ```suggestion
## Using persistent volumes in a pod
```

> Once you have a Persistent Volume Claim you can [claim it as a Volume](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#claims-as-volumes) in your Pods.

Reviewer: ```suggestion
Once you have a PersistentVolumeClaim you can [claim it as a volume](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#claims-as-volumes) in your pods.
```

> Note that an EBS Volume can only be used by a single Pod at the same time. Thus, the access mode of your PVC can only be `ReadWriteOnce`.

Reviewer: ```suggestion
Note that an EBS volume can only be used by a single pod at the same time. Thus, the access mode of your PVC can only be `ReadWriteOnce`.
```

> Under the hood the EBS Volume stays detached from your nodes as long as it is not claimed by a Pod. As soon as a Pod claims it, it gets attached to the node that holds the Pod.

Reviewer: ```suggestion
Under the hood the EBS volume stays detached from your nodes as long as it is not claimed by a pod. As soon as a pod claims it, it gets attached to the EC2 instance of the node that holds the pod.
```

## Example

First we create a PVC:

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: my-claim
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: ebs-csi
  resources:
    requests:

>       storage: 7Gi

Author: ```suggestion
      storage: 6Gi
```

```

> Note that the Access Mode while being fixed with EBS still needs to be defined as `ReadWriteOnce` in the manifest.

Reviewer: ```suggestion
Note that the access mode, while being fixed with EBS, still needs to be defined as `ReadWriteOnce` in the manifest (see the `accessModes` property).
```

> Further, we need to define the Storage Class `ebs-csi` because it is not the default storage class (default `gp2`).

Reviewer: ```suggestion
Further, we need to define the storage class `ebs-csi` because it is not the default storage class (default `gp2`).
```

> Further, we need to define the Storage Class `ebs-csi` because it is not the default storage class (default `gp2`) the Kubernetes cluster will just take the default storage class (here `gp2`).

Author: ```suggestion
Further, we need to define the Storage Class `ebs-csi` because it is not the default storage class (default `gp2`).
```

Now we can create a Pod that uses our PVC:

```yaml
kind: Pod
apiVersion: v1
metadata:
  name: mypod
spec:
  containers:
    - name: myfrontend
      image: nginx
      volumeMounts:
      - mountPath: "/var/www/html"
        name: mypd
  volumes:
    - name: mypd
      persistentVolumeClaim:
        claimName: my-claim
```

> Now we have an NGINX Pod which serves the contents of our EBS Volume.

Reviewer: ```suggestion
Now we have an NGINX pod which serves the content of our EBS volume.
```

> ## Expanding Persistent Volume Claims

Reviewer: Can we change this headline to a more intuitive one, closer to the user goal in mind? Users would likey think about "increasing volume size" or "increasing storage capacity of a volume". The fact that this is done via the PVC is more like a detail of the solution.

> Starting with Kubernetes v1.16 CSI Volume can be expanded by simply editing the claim and requesting a larger size. It will trigger an update in the underlying Persistent Volume and EBS Volume (Kubernetes always uses the existing one).

Reviewer: ```suggestion
Starting with Kubernetes v1.16, CSI volumes can be expanded by simply editing the claim (the `PersistentVolumeClaim` resource) and requesting a larger size. This will trigger an update of the associated `PersistentVolume` resource and the underlying EBS volume.
```

I think `(Kubernetes always uses the existing one)` is expected and thus not necessary.

> Resizing an in-use PersistentVolumeClaim is enabled by default with Kubernetes v1.15, there is no need to delete and recreate a Pod. You can only resize volumes containing a file system if the file system is XFS, Ext3, or Ext4.

Reviewer: ```suggestion
Resizing a PersistentVolumeClaim resource in use is enabled by default with Kubernetes v1.15, so there is no need to delete and recreate a pod.

**Note:** You can only resize volumes formatted with one of the file systems XFS, Ext3, or Ext4.
```

> __Warning__: Expanding EBS volumes is a time consuming operation. Also, there is a per-volume quota of one modification every 6 hours.

Reviewer: ```suggestion
__Warning__: Expanding EBS volumes is a time-consuming operation. Also, there is a per-volume quota of 1 modification every 6 hours.
```

> ## Deleting Persistent Volumes

Reviewer: ```suggestion
## Deleting persistent volumes
```

> The Reclaim Policy is set to `Delete`. Thus, deleting the `PersistentVolume` resource will also delete the respective EBS Volume. Similarly, by default if you delete a `PersistentVolumeClaim` resource the respective Persistent Volume and EBS will get deleted.

Reviewer: What is the Reclaim Policy? I haven't seen it mentioned so far. Ideally explain it briefly in a sentence before this one.

```suggestion
The Reclaim Policy is set to `Delete`. Thus, deleting the `PersistentVolume` resource will also delete the respective EBS volume. Similarly, by default if you delete a `PersistentVolumeClaim` resource, the respective `PersistentVolume` and EBS volume will get deleted.
```

> The Reclaim Policy is set to `Delete`. Thus, deleting the `PersistentVolume` resource will also delete the respective EBS Volume. Similarly, by default if you delete a `PersistentVolumeClaim ` resource the respective Persistent Volume and EBS will get deleted.

Author: ```suggestion
The Reclaim Policy is set to `Delete`. Thus, deleting the `PersistentVolume` resource will also delete the respective EBS Volume. Similarly, by default if you delete a `PersistentVolumeClaim` resource the respective Persistent Volume and EBS will get deleted.
```

> The Reclaim Policy is set to `Delete`. Thus, deleting the `PersistentVolume` resource will also delete the respective EBS Volume. Similarly, by default if you delete a `PersistenVolumeClaim` resource the respective Persistent Volume and EBS will get deleted.

Author: ```suggestion
The Reclaim Policy is set to `Delete`. Thus, deleting the `PersistentVolume` resource will also delete the respective EBS Volume. Similarly, by default if you delete a `PersistentVolumeClaim ` resource the respective Persistent Volume and EBS will get deleted.
```

> Note that deleting an application that is using Persistent Volumes might not automatically also delete its storage, so in most cases you will have to manually delete the `PersistentVolumeClaim` resources to clean up.

Reviewer: ```suggestion
Note that deleting an application that is using persistent volumes might not automatically also delete its storage, so in most cases you will have to manually delete the `PersistentVolumeClaim` resources to clean up.
```

The term "application" is not really defined well, I think. Before we used "pod". An alternative could also be "workload".

> Note that deleting an application that is using Persistent Volumes might not automatically also delete its storage, so in most cases you will have to manually delete the `PersistenVolumeClaim` resources to clean up.

Author: ```suggestion
Note that deleting an application that is using Persistent Volumes might not automatically also delete its storage, so in most cases you will have to manually delete the `PersistentVolumeClaim` resources to clean up.
```

## Further reading

> - [AWS Storage Classes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#aws)

Reviewer: ```suggestion
- [AWS storage classes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#aws)
```

> - [Persistent Volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistent-volumes)

Reviewer: ```suggestion
- [Persistent volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistent-volumes)
```

> - [Persistent Volume Claims](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims)

Reviewer: ```suggestion
- [Persistent volume claims](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims)
```

> - [Claim Persistent Volumes in Pods](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#claims-as-volumes)

Reviewer: ```suggestion
- [Claim persistent volumes in pods](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#claims-as-volumes)
```

> - [Kubernetes CSI Developer Documentation](https://kubernetes-csi.github.io/docs/)

Reviewer: ```suggestion
- [Kubernetes CSI developer documentation](https://kubernetes-csi.github.io/docs/)
```

The aws-ebs-csi-driver-app, offered via our [App Platform](/basics/app-platform/) in the Giant Swarm Playground catalog, provides a CSI interface to manage the lifecycle of Amazon EBS volumes.

## Installing the EBS CSI driver

To install the EBS CSI driver you will need to follow these steps:

1. Access the [web interface](/reference/web-interface/) and select the cluster on which you want to install the EBS CSI driver.
2. Open the Apps tab.
3. Click the Install App button
4. Select the Giant Swarm Playground catalog.
5. Select the App named aws-ebs-csi-driver-app.
6. Click the Configure & Install button. Make sure that the correct cluster is selected.
7. Click the Install App button.

## Storage classes

Once you installed the `aws-ebs-csi-driver-app` from _Giant Swarm Playground_ catalog, your Kubernetes cluster will have a Storage Class `ebs-csi` deployed.

The storage class offers volume encryption and allows resizing by default.

## Creating persistent volumes

> The most straight forward way to create a persistent volume is to create a [PersistentVolumeClaim](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims) resource, which will automatically create a corresponding `PersistentVolume` (PV) resource for you.

Reviewer 2: might be worth mentioning also here that you need to configure storage class on that PVC< it can be seen in the example below but its not explicitly said here which might be confusing

First we create a PVC:

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: my-claim
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: ebs-csi
  resources:
    requests:
      storage: 5Gi
```

Note that the access mode, while being fixed with EBS, still needs to be defined as `ReadWriteOnce` in the manifest (see the `accessModes` property).

Further, we need to define the storage class `ebs-csi` because it is not the default storage class (default `gp2`).

Now we can create a Pod that uses our PVC:

```yaml
kind: Pod
apiVersion: v1
metadata:
  name: mypod
spec:
  containers:
    - name: myfrontend
      image: nginx
      volumeMounts:
      - mountPath: "/var/www/html"
        name: mypd
  volumes:
    - name: mypd
      persistentVolumeClaim:
        claimName: my-claim
```

Now we have an NGINX pod which serves the content of our EBS volume.

## Increasing storage capacity of a volume

Starting with Kubernetes v1.16, CSI volumes can be expanded by simply editing the claim (the `PersistentVolumeClaim` resource) and requesting a larger size. This will trigger an update of the associated `PersistentVolume` resource and the underlying EBS volume.

> Resizing a PersistentVolumeClaim resource in use is enabled by default with Kubernetes v1.15, so there is no need to delete and recreate a pod.

Reviewer: ```suggestion
Resizing a `PersistentVolumeClaim` resource in use is enabled by default with Kubernetes v1.15, so there is no need to delete and recreate a pod.
```

**Note:** You can only resize volumes formatted with one of the file systems XFS, Ext3, or Ext4.

> __Warning__: Expanding EBS volumes is a time-consuming operation. Also, there is a per-volume quota of 1 modification every 6 hours.

Reviewer: ```suggestion
**Warning:** Expanding EBS volumes is a time-consuming operation. Also, there is a per-volume quota of 1 modification every 6 hours.
```

Consistency nitpick
